colcon_ros_domain_id_coordinator/shell/ros_domain_id.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 Open Source Robotics Foundation, Inc.
+# Licensed under the Apache License, Version 2.0
+
+from asyncio import Queue
+from asyncio import QueueEmpty
+import os
+from pathlib import Path
+from random import shuffle
+
+from colcon_core.plugin_system import satisfies_version
+from colcon_core.plugin_system import SkipExtensionException
+from colcon_core.shell import logger
+from colcon_core.shell import ShellExtensionPoint
+
+
+class ROSDomainIDShell(ShellExtensionPoint):
+    """Set a different ROS_DOMAIN_ID environment variable for each task."""
+
+    # The priority should be higher than any usable shells
+    PRIORITY = 900
+
+    def __init__(self):  # noqa: D107
+        super().__init__()
+        satisfies_version(ShellExtensionPoint.EXTENSION_POINT_VERSION, '^2.2')
+
+        self.free_ids = Queue()
+
+        all_ids = []
+        all_ids.extend(range(1, 102))
+        all_ids.extend(range(215, 233))
+        shuffle(all_ids)
+        for i in all_ids:
+            self.free_ids.put_nowait(str(i))
+
+    def get_file_extensions(self):  # noqa: D102
+        return ()
+
+    def create_prefix_script(self, prefix_path, merge_install):  # noqa: D102
+        return []
+
+    def create_package_script(  # noqa: D102
+        self, prefix_path, pkg_name, hooks
+    ):
+        return []
+
+    def create_hook_set_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, value,
+    ):
+        return Path('ros_domain_id')
+
+    def create_hook_append_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, subdirectory,
+    ):
+        return Path('ros_domain_id')
+
+    def create_hook_prepend_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, subdirectory,
+    ):
+        return Path('ros_domain_id')
+
+    async def generate_command_environment(  # noqa: D102
+        self, task_name, build_base, dependencies,
+    ):
+        try:
+            task_id = self.free_ids.get_nowait()
```

Comment:
These aren't really `task_id`s that get returned, they are `domain_id`s, correct?  I think it would be a little clearer if the variable was renamed.

Reply:
By `task_id` I meant "the domain ID for this task". The nomenclature is really a holdover from when I was trying to assign the ID to the task and record it as such, and it just looks confusing now.

I've changed it to `domain_id`.